Rework rmake support library API

jieyouxu opened the pull request:
### Take 1: Strongly-typed API

Context: https://github.com/rust-lang/rust/pull/122448#discussion_r1523774427

> My 2 cents: from my experience with writing similar "test DSLs", I would suggest to create these helpers as soon as possible in the process (basically the first time someone needs them, not only after N similar usages), and basically treat any imperative code in these high-level tests as a maintenance burden, basically making them as declarative as possible. Otherwise it might be a bit annoying to keep refactoring the tests later once such helpers are available.
>
> I would even discourage the arg method and create explicit methods for setting things like unpretty, the output file etc., but this might be more controversial, as it will make the invoked command-line arguments more opaque.

cc @Kobzol for the testing DSL suggestion.

Example:

```rs
let output = Rustc::new()
    .input_file("main.rs")
    .emit(&[EmitKind::Metadata])
    .extern_("stable", &stable_path)
    .output();
```

### Take 2: xshell-based macro API

Example:

```rs
let sh = Shell::new()?;
let stable_path = stable_path.to_string_lossy();
let output = cmd!(sh, "rustc main.rs --emit=metadata --extern stable={stable_path}").output()?;
```

### Take 3: Weakly-typed API with a few helper methods

```rs
let output = Rustc::new()
    .input("main.rs")
    .emit("metadata")
    .extern_("stable", &stable_path)
    .output();
```

jieyouxu committed:
Rework rmake support library to have a weakly-typed API with helper methods

Nilstrieb commented on tests/run-make/a-b-a-linker-guard/rmake.rs:
```diff
@@ -1,44 +1,37 @@
+        .input_file("b.rs")
+        .enable_unstable_options()
+        .codegen_opt(PreferDynamic)
+        .codegen_opt(SymbolManglingVersion(SymbolManglingVersion::Legacy))
```
I'm not sure whether this really is better than what we had before (with merging the two like `-Cprefer-dynamic` into one line, which could be asserted by arg panicking when it sees -Z or -C).

jieyouxu replied:
Don't really have a super strong opinion for the support library API atm, still experimenting *shrug*

It does hardcode the cli interface into the support library. But if the cli interface changes, then surely the tests affected will have to change with it anyway?

jieyouxu replied:
Also I can omit the codegen_opt layer and just provide e.g. `prefer_dynamic()` / `lto(LtoKind::Fat)` directly.

Nilstrieb replied:
I think arg is still simpler, easier to maintain and more obvious to understand

Nilstrieb commented on tests/run-make/a-b-a-linker-guard/rmake.rs:
```diff
@@ -1,44 +1,37 @@
-        .arg("symbol-mangling-version=legacy")
+    Rustc::new()
+        .input_file("a.rs")
+        .cfg("x")
```
I do like this helper, it makes the test shorter

klensy commented:
Before: to write test you needed only to write almost as if bash command lines.
After: now you need to learn that DSL.

Maybe something like https://docs.rs/xshell/latest/xshell/ ? Used in r-a and looks cute.

Kobzol commented:
Oh, I see you really went for it full steam ahead :laughing: This approach is of course highly bike-sheddable and I'm sure that it will have both fans and people who don't like it.

Now that I look at it, I have to agree with Nilstrieb [here](https://github.com/rust-lang/rust/pull/122448#discussion_r1524358152) that perhaps the strongly typed wrappers are too much, and they make it unnecessarily opaque to the reader - who probably knows the CLI flag, but might not know the wrapper. It's true that with the wrapper we can easily search for all usages of that flag in the run-make tests, but if you want to grep e.g. for all usages of `-Zunpretty=hir` (across all tests), then this actually makes that harder. So perhaps using the raw flags is a better choice after all.

On the other hand, I would keep some of the high-level helpers for things that are super common, but not very interesting and are "almost always boilerpate", for example the output `-o` parameter configuration.

Nilstrieb commented:
i agree that common things like -o deserve helpers, especially (or maybe rather *only*) if they make the code shorter/simpler.
another thing I'd suggest is making `arg` panic when it sees plain `-C` or `-Z` to force people to use `-Za` form instead of `-Z a` (as it makes the code significantly shorter and simpler).

jieyouxu commented:
> Before: to write test you needed only to write almost as if bash command lines. After: now you need to learn that DSL.
> 
> Maybe something like https://docs.rs/xshell/latest/xshell/ ? Used in r-a and looks cute.

I'm not sure why I thought it was a problem when I did the run-make infra PR initially, when I could've just re-exported xshell.

Now that I did write the super strongly-typed API, I think the shell version is just easier to read lol

```rs
extern run_make_support;
use run_make_support::anyhow;
use run_make_support::xshell::{Shell, cmd};

fn main() -> anyhow::Result<()> {
  let sh = Shell::new()?;
  let _blah = cmd!(sh, "rustc a.rs -Zunpretty=hir").read()?;
}
```

Maybe just go with the shell version after all? I want to get the API bikeshedding done before other tests are ported over :3

Kobzol commented:
That approach could make it simpler to port the existing tests.

I would still keep some "context object", on which we can either call `arg` etc. for more complicated stuff, or just run the shell for simple use-cases. I'd definitely hide the usage of `Shell` behind our own API, to avoid exposing it directly to the test - if that is possible to do w.r.t. the macro.

In test DSLs, I often appreciate how I can do asserts on the return value of something, e.g.
```rust
let res = Command::new().set_arg(foo).run().assert_ok();
assert_eq!(res.get_output("bar"), "baz");
```
But I'm not actually sure if we need that in runmake tests, since we usually just invoke the compiler and then check the filesystem.

One thing that could be annoying with the shell approach is setting default values. E.g., we might want to set `-o` by default, or the `rustc` path by default, or redirect stdout/stderr of the compiler to some file, upon which we later want to do asserts etc. We should keep this option if we decide to use the shell. Replacing a Makefile with a Rust file that runs a fully unstructured and opaque shell command isn't such an improvement, I'd say :) We want to get rid of the shell-ism where possible.

jieyouxu commented:
So the xshell version for a-b-a-linker-guard looks something like:

```rs
fn main() -> anyhow::Result<()> {
    let sh = Shell::new()?;
    let rustc = rustc();
    let out_dir = out_dir();
    cmd!(
        sh,
        "{rustc} a.rs --out-dir={out_dir} -L {out_dir} --cfg x -C prefer-dynamic \
        -Z unstable-options -C symbol-mangling-version=legacy"
    )
    .run()?;
    cmd!(
        sh,
        "{rustc} b.rs --out-dir={out_dir} -L {out_dir} -C prefer-dynamic -Z unstable-options \
        -C symbol-mangling-version=legacy"
    )
    .run()?;

    run("b");

    cmd!(
        sh,
        "{rustc} a.rs --out-dir={out_dir} -L {out_dir} --cfg y -C prefer-dynamic \
        -Z unstable-options -C symbol-mangling-version=legacy"
    )
    .run()?;

    run_fail("b");
}
```

jieyouxu commented:
Now that I see how this version looks like... I think I personally prefer the typed API version, but with only a selected few helper methods (and not full-blown typed API). That way, we can still have presets (like `aux_build` et al.) and don't have to deal with a dependency on xshell.

jieyouxu commented:
With the typed-but-not-too-stringly API:

```rs
fn main() {
    Rustc::new()
        .input("a.rs")
        .cfg("x")
        .arg("-Zunstable-options")
        .arg("-Cprefer-dynamic")
        .arg("-Csymbol-mangling-version=legacy")
        .run();

    Rustc::new()
        .input("b.rs")
        .arg("-Zunstable-options")
        .arg("-Cprefer-dynamic")
        .arg("-Csymbol-mangling-version=legacy")
        .run();

    run("b");

    Rustc::new()
        .input("a.rs")
        .cfg("y")
        .arg("-Zunstable-options")
        .arg("-Cprefer-dynamic")
        .arg("-Csymbol-mangling-version=legacy")
        .run();

    run_fail("b");
}
```

Kobzol commented:
That looks like a good compromise to me. `cfg` will be less greppable due to this though (which is true basically for any helper method that we introduce on the rustc wrapper), but I don't think that a big issue for this specific flag.

workingjubilee commented:
@rustbot ready

wait...

Nilstrieb commented on tests/run-make/wasm-custom-section/rmake.rs:
```diff
@@ -1,15 +1,16 @@
 
-    rustc().arg("foo.rs").arg("--target=wasm32-wasip1").run();
-    rustc().arg("bar.rs").arg("--target=wasm32-wasip1").arg("-Clto").arg("-O").run();
+    Rustc::new().input("foo.rs").target("wasm32-wasip1").run();
```
Why `rustc()`->`Rustc::new()`?

jieyouxu replied:
Previously `rustc()` just called `RustcInvocationBuilder::new()` (which is now `Rustc`), it felt like an really unnecessary wrapper

Nilstrieb replied:
I think it looks nicer

Nilstrieb commented:
I'm going to approve this, we can always change it later.
@bors r+

bors commented:
:pushpin: Commit 6ad3b4bb85524a4245ac6e994ac438d555cc1cd5 has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb 6ad3b4bb85524a4245ac6e994ac438d555cc1cd5 -->
<!-- homu: {"type":"Approved","sha":"6ad3b4bb85524a4245ac6e994ac438d555cc1cd5","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Nilstrieb commented:
@jieyouxu asked me to @bors r-

jieyouxu commented:
@rustbot ready

jieyouxu commented:
Changes since last reviewed:
- Reintroduced `rustc` and `aux_build` helper functions, and updated tests accordingly.

Nilstrieb commented:
@bors r+ r+ r+ r+ r+

bors commented:
:pushpin: Commit c828984dcaa1495ccaf3cfbc842cbc8ddd076bca has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb c828984dcaa1495ccaf3cfbc842cbc8ddd076bca -->
<!-- homu: {"type":"Approved","sha":"c828984dcaa1495ccaf3cfbc842cbc8ddd076bca","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:pushpin: Commit c828984dcaa1495ccaf3cfbc842cbc8ddd076bca has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb c828984dcaa1495ccaf3cfbc842cbc8ddd076bca -->
<!-- homu: {"type":"Approved","sha":"c828984dcaa1495ccaf3cfbc842cbc8ddd076bca","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #122849

bors commented:
:pushpin: Commit c828984dcaa1495ccaf3cfbc842cbc8ddd076bca has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb c828984dcaa1495ccaf3cfbc842cbc8ddd076bca -->
<!-- homu: {"type":"Approved","sha":"c828984dcaa1495ccaf3cfbc842cbc8ddd076bca","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #122849

bors commented:
:pushpin: Commit c828984dcaa1495ccaf3cfbc842cbc8ddd076bca has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb c828984dcaa1495ccaf3cfbc842cbc8ddd076bca -->
<!-- homu: {"type":"Approved","sha":"c828984dcaa1495ccaf3cfbc842cbc8ddd076bca","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #122849

bors commented:
:pushpin: Commit c828984dcaa1495ccaf3cfbc842cbc8ddd076bca has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb c828984dcaa1495ccaf3cfbc842cbc8ddd076bca -->
<!-- homu: {"type":"Approved","sha":"c828984dcaa1495ccaf3cfbc842cbc8ddd076bca","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Nilstrieb commented:
oh no

matthiaskrgr commented:
@bors r- r- r- r-  rollup=iffy
https://github.com/rust-lang/rust/pull/122932#issuecomment-2016502323

jieyouxu commented:
I accidentally deleted a `-Clto` argument in that test. Oops.

jieyouxu commented:
@rustbot ready

Nilstrieb commented:
@bora r+ (just a single one this time)

Nilstrieb commented:
@bors r+ fuck

bors commented:
:pushpin: Commit 4e8753b26d8203feb10164a8d212967870e4bcb1 has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb 4e8753b26d8203feb10164a8d212967870e4bcb1 -->
<!-- homu: {"type":"Approved","sha":"4e8753b26d8203feb10164a8d212967870e4bcb1","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout rmake-example-refactor` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self rmake-example-refactor --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging src/tools/run-make-support/src/lib.rs
CONFLICT (content): Merge conflict in src/tools/run-make-support/src/lib.rs
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #122966) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jieyouxu commented:
Changes since last reviewed:
- Fixed merge conflicts.
- Pull out rustc and rustdoc to their own modules.
- Unshare setup logic because rustdoc should not have `--out-dir` set by default (run-make rustdoc tests seem to usually need to specify `--out-dir` themselves, and specifying `--out-dir` already would prevent overriding).
- Introduce `out_dir`, `input` and `arg_file` helper methods on `Rustdoc`.

Nilstrieb commented:
@bors r+

bors commented:
:hourglass: Testing commit 1f2178b9e7b43f4efe549e19889eafeec9061e8c with merge 0157da41ee9a1f6fb1a20e00ea62ab4122e2feaa...
<!-- homu: {"type":"BuildStarted","head_sha":"1f2178b9e7b43f4efe549e19889eafeec9061e8c","merge_sha":"0157da41ee9a1f6fb1a20e00ea62ab4122e2feaa"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8456250849/job/23170748200)
Approved by: Nilstrieb
Pushing 0157da41ee9a1f6fb1a20e00ea62ab4122e2feaa to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Nilstrieb","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8456250849/job/23170748200"},"merge_sha":"0157da41ee9a1f6fb1a20e00ea62ab4122e2feaa"} -->

rust-timer commented:
Finished benchmarking commit (0157da41ee9a1f6fb1a20e00ea62ab4122e2feaa): [comparison URL](https://perf.rust-lang.org/compare.html?start=d5db7fb537c0352eec3855aa0331c271c48ac4f6&end=0157da41ee9a1f6fb1a20e00ea62ab4122e2feaa&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d5db7fb537c0352eec3855aa0331c271c48ac4f6&end=0157da41ee9a1f6fb1a20e00ea62ab4122e2feaa&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.6% | [1.3%, 6.0%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d5db7fb537c0352eec3855aa0331c271c48ac4f6&end=0157da41ee9a1f6fb1a20e00ea62ab4122e2feaa&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.6% | [0.4%, 2.2%] | 11    |
| Regressions ❌ <br /> (secondary)  | 3.8% | [1.9%, 6.6%] | 15    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.6% | [0.4%, 2.2%] | 11    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d5db7fb537c0352eec3855aa0331c271c48ac4f6&end=0157da41ee9a1f6fb1a20e00ea62ab4122e2feaa&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.9%  | [1.9%, 1.9%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 4.2%  | [4.2%, 4.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.9%, -0.5%] | 10    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.4% | [-0.9%, 1.9%]  | 11    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d5db7fb537c0352eec3855aa0331c271c48ac4f6&end=0157da41ee9a1f6fb1a20e00ea62ab4122e2feaa&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 670.379s -> 671.34s (0.14%)
**Artifact size**: 315.75 MiB -> 315.76 MiB (0.00%)

Kobzol commented:
This only modified tests, so the result is noise.

 @rustbot label: +perf-regression-triaged
